# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Shape.cpp
# Commit: 46f7a204fe42
# Full Hash: 46f7a204fe4211862457439fd2e4d26b39f3d5b4
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-05-18 21:36:28
# Description:
#   Bug 1703481 - Ensure proto object has a unique id when setting the IsUsedAsPrototype flag. r=jonco
#   
#   This way we only do the duplicate lookup the first time an object is used as prototype.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D114076
# ==============================================================================

diff -r 68062ca46e90 -r 46f7a204fe42 js/src/vm/Shape.cpp
--- a/js/src/vm/Shape.cpp	Tue May 18 08:33:20 2021 +0000
+++ b/js/src/vm/Shape.cpp	Tue May 18 08:57:34 2021 +0000
@@ -1739,6 +1739,12 @@
     if (!JSObject::setIsUsedAsPrototype(cx, protoObj)) {
       return nullptr;
     }
+    // Ensure the proto object has a unique id to prevent OOM crashes below.
+    uint64_t unused;
+    if (!cx->zone()->getOrCreateUniqueId(protoObj, &unused)) {
+      ReportOutOfMemory(cx);
+      return nullptr;
+    }
     proto = TaggedProto(protoObj);
   }
 
